src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -184,10 +184,22 @@ double FFN<OutputLayerType, InitializationRuleType>::Evaluate(
   currentTarget = responses.unsafe_col(i);
 
   Forward(std::move(currentInput));
-  double res = outputLayer.Forward(std::move(boost::apply_visitor(
+  currentCost = outputLayer.Forward(std::move(boost::apply_visitor(
       outputParameterVisitor, network.back())), std::move(currentTarget));
 
-  return res;
+  return currentCost;
+}
+
+template<typename OutputLayerType, typename InitializationRuleType>
+double FFN<OutputLayerType, InitializationRuleType>::Evaluate(
```

Comment:
Currently, if the optimizer calls the Gradient function we do Evaluate -> Backward -> Gradient inside the FFN class. Afterward, the Optimizer calls the Evaluate function again to keep track of the the latest changes (did the performance improve).

But with this change the performance analysis, the Evaluation call inside the Optimizer after the Gradient returns the same value as before, which is problematic if the input is just a single element since SGD will return with:
```
if (std::abs(lastObjective - overallObjective) < tolerance)
{
  Log::Info << "SGD: minimized within tolerance " << tolerance << "; "
      << "terminating optimization." << std::endl;
  return overallObjective;
}
```
The question is, do we handle that situation or do we expect samples > 1?